README.md
# Netronic

[[**Homepage**](https://ed7n.github.io/netronic)]

## Building

    $ src/scss/build.sh && src/xfwm4/build.sh && cp -ruv src/ABOUT src/SETUP src/qt5ct src/wine release/netronic


CHANGES
Netronic: Changes

—Update 1

——Revision 7, 12/21/2023.
- WinRT theme registry values.
- Swapped Wine light and dark colors in Wine dark scheme.

——Revision 6, 04/10/2023.
- Explicit underlining of links.
- Minimum metrices for GTK buttons.
- Removed check and radio icon scaling.
- Removed troubleshooting of icons from setup manual.
- SCSS root-relative pixel spacer.
  - Eliminated need for DPI variations on GTK themes.
- Updated SCSS variables.

——Revision 5, 04/04/2023.
- Completed build command in README.
- GTK 4 menu button arrows.
- GTK 4 rows in dropdowns workaround.
- GTK 4 statusbar styles.
  - Statusbars exist in GTK 3 but not needed to be styled.
- Variablized SCSS executable.

——Revision 4, 03/27/2023.
- Initial GTK 4 support.
- Removed Handy dark scheme workaround.
  - Provided instructions instead.

——Revision 3, 03/23/2023.
- Handy dark scheme workaround.
- Updated SCSS variables.

——Revision 2, 12/31/2022.
- Project header in `ABOUT`.
- Rebuilt GTK 3 stylesheets with Sass 1.57.1.
- Refactored build scripts.
- Renamed `VERSION` to `ABOUT`.
- Updated setup manual.

——Revision 1, 07/27/2022.
- Fixed xfwm4 dark inactive window border colors.

——Revision 0, 07/25/2022.
- Build scripts.
- `VERSION` file.
- Win9 color scheme.
- Wine control alternate background colors.
- Xfdesktop colors.
- XHDPI (200%) themes.
-- XHDPI (150%) is renamed to HDPI.
- XXHDPI themes.
- Edited `calendar:indeterminate` color.
- Edited HDPI xfwm4 close button.
- Edited package hierarchy.
- Edited `scrollbar.overlay-indicator` color and opacity.
- Edited setup manual.
- Fixed `.app-notification` border and padding.
- Fixed `treeview.view check` and `treeview.view radio` background colors.
- Isolated xfwm4 text colors from GTK 3.
- Moved GTK colors to Sass variables.
- Omitted unneeded comments from outputs.
- Renamed HDPI to MHDPI, and XHDPI to HDPI.
- Renamed Qt5 color schemes.
- Removed `index.theme` metadata files.
- Removed unused dependencies on `sass:math`.
- Variablized Thunar `treeview.view` outline offset.
- Variablized xfce-notify-4.0 stylesheets.

—Update 0

——Revision 7, 06/12/2022.
- Xfwm4 horizontal title offset.

——Revision 6, 05/15/2022.
- Colors-only Qt5 port.
- Epiphany preferences `viewswitcher` button workaround.
- XHDPI themes.
- Edited (X)HDPI naming.
- Fixed dark-hdpi menu-prelight and -pressed colors.
- Restored Adwaita's way of spacing `check`, `checkbutton`, `radio`, and
  `radiobutton`.
- Restored `.text-button` to `checkbutton` and `radiobutton` selectors except
  for hover highlights.
- Updated setup manual for magenta placeholders.

——Revision 5, 05/03/2022.
- Removed `.text-button` from `checkbutton` and `radiobutton` selectors.
-- [https://gitlab.xfce.org/apps/mousepad/-/issues/169]
- Updated setup manual for AppArmor, Firefox Proton, and Wine theming.

——Revision 4, 02/28/2022.
- `.content-view` background.

——Revision 3, 02/26/2022.
- Epiphany webpage tab button styles.

——Revision 2, 11/06/2021.
- Fixed button backdrop disabled colors.

——Revision 1, 11/04/2021.
- Fixed Wine selection background color.
- Tweaked selection background color.

——Revision 0, 10/19/2021.
- Initial release.


src/xfwm4/build-xfwm4.sh
#!/bin/env bash

{
  declare -p ews || declare -Ax ews=([base]="${0%/*}" [exec]="${0}" \
      [name]='Netronic Xfwm4 Theme Builder' [sign]='by Brendon, 07/25/2022.')
} &> /dev/null

# Base output filename.
ews[bon]='Netronic'
# Base output path.
ews[bop]="${ews[base]}"'/../../release/netronic/themes'
# Primary DPI.
ews[ddi]='mdpi'
# Primary variant.
ews[dva]='light'
# Secondary DPIs and variants.
declare -ar NTB_DPIS=('mhdpi' 'hdpi' 'xhdpi' 'xxhdpi') \
    NTB_VARS=('dark' 'win9')

run() {
  local ntbInp="${1}" ntbOut="${2}"
  shift 2
  echo -en "${ntbOut}"'\r'
  . "${ews[inp]}"/"${ntbInp}".sh "${@}" > "${ews[out]}"/"${ntbOut}" || exit
  echo -en '\r'"${ntbOut//?/ }"'\r'
}

wrt() {
  local ntbStu
  echo 'Writing `'"${ews[var]}"'-'"${ews[dpi]}"'`.'
  mkdir -p "${ews[out]}" || exit
  for ntbSte in 'active' 'inactive'; do
    ntbStu="${ntbSte@u}"
    local -n \
        ntbCtb='colorWindow'"${ntbStu}"'TitlebarBackground' \
        ntbCbi='colorWindow'"${ntbStu}"'BorderTitle' \
        ntbCbt='colorWindow'"${ntbStu}"'BorderTop' \
        ntbCbb='colorWindow'"${ntbStu}"'BorderBottom'
    for ntbNum in '1' '2' '3' '4' '5'; do
      run 'top' 'title-'"${ntbNum}"'-'"${ntbSte}"'.xpm' "${ntbCtb}" \
          "${ntbCbi}" "${ntbCbt}"
    done
    run 'bottom' 'bottom-'"${ntbSte}"'.xpm' "${ntbCbb}"
    for ntbSde in 'left' 'right'; do
      local -n ntbCsd='colorWindow'"${ntbStu}"'Border'"${ntbSde@u}"
      run 'sides' "${ntbSde}"'-'"${ntbSte}"'.xpm' "${ntbCsd}"
      run 'top''-'"${ntbSde}" 'top-'"${ntbSde}"'-'"${ntbSte}"'.xpm' \
          "${ntbCtb}" "${ntbCbi}" "${ntbCbt}" "${ntbCsd}"
      run 'bottom-'"${ntbSde}" 'bottom-'"${ntbSde}"'-'"${ntbSte}"'.xpm' \
          "${ntbCbb}" "${ntbCsd}"
    done
  done
  for ntbSte in 'active' 'inactive' 'prelight' 'pressed'; do
    ntbStu="${ntbSte@u}"
    local -n \
        ntbCfg='colorButton'"${ntbStu}" \
        ntbCbg='colorButton'"${ntbStu}"'Background' \
        ntbCbt='colorButton'"${ntbStu}"'BorderTop' \
        ntbCbl='colorButton'"${ntbStu}"'BorderLeft' \
        ntbCbr='colorButton'"${ntbStu}"'BorderRight' \
        ntbCbb='colorButton'"${ntbStu}"'BorderBottom'
    for ntbBtn in 'close' 'hide' 'maximize' 'shade' 'stick'; do
      run "${ntbBtn}" "${ntbBtn}"'-'"${ntbSte}"'.xpm' "${ntbCfg}" "${ntbCbg}" \
          "${ntbCbt}" "${ntbCbl}" "${ntbCbr}" "${ntbCbb}"
    done
    case "${ntbSte}" in
      'prelight' | 'pressed' )
        local ntbCfg="${ntbCbg}" ;;&
      'active' | 'prelight' | 'pressed' )
        local ntbCbg="${colorWindowActiveTitlebarBackground}" ;;
      'inactive' )
        local ntbCbg="${colorWindowInactiveTitlebarBackground}" ;;
      * )
        echo 'Bad ${ntbSte}: '"${ntbSte}"
        exit 1 ;;
    esac
    run 'menu' 'menu-'"${ntbSte}"'.xpm' "${ntbCfg}" "${ntbCbg}"
    local -n \
        ntbCfg='colorButtonToggled'"${ntbStu}" \
        ntbCbg='colorButtonToggled'"${ntbStu}"'Background' \
        ntbCbt='colorButtonToggled'"${ntbStu}"'BorderTop' \
        ntbCbl='colorButtonToggled'"${ntbStu}"'BorderLeft' \
        ntbCbr='colorButtonToggled'"${ntbStu}"'BorderRight' \
        ntbCbb='colorButtonToggled'"${ntbStu}"'BorderBottom'
    for ntbBtn in 'maximize' 'shade' 'stick'; do
      run "${ntbBtn}" "${ntbBtn}"'-toggled-'"${ntbSte}"'.xpm' "${ntbCfg}" \
          "${ntbCbg}" "${ntbCbt}" "${ntbCbl}" "${ntbCbr}" "${ntbCbb}"
    done
  done
  run 'themerc' 'themerc' "${colorWindowActiveTitlebar}" \
      "${colorWindowInactiveTitlebar}"
}

echo -e "${ews[name]}"' '"${ews[sign]}"'\n'
ews[bop]="${ews[bop]%/}"
IFS=' '
for ntbVar in '' ${NTB_VARS[*]}; do
  ews[var]="${ntbVar:-${ews[dva]}}"
  echo 'Reading `'"${ews[var]}"'`.'
  . "${ews[base]}"/color/"${ews[var]}".sh || exit
  for ntbDpi in '' ${NTB_DPIS[*]}; do
    ews[dpi]="${ntbDpi:-${ews[ddi]}}"
    ews[inp]="${ews[base]}"'/dpi/'"${ews[dpi]}"
    ews[out]=''
    [ "${ntbDpi}" ] && ews[out]="${ews[out]}"'-'"${ntbDpi}"
    [ "${ntbVar}" ] && ews[out]="${ews[out]}"'-'"${ntbVar}"
    ews[out]="${ews[bop]}"'/'"${ews[bon]}""${ews[out]}"'/xfwm4'
    wrt
  done
done
echo 'Done.'
